File: Src/static/js/perfilesAsociarComp.js
function verInstructores(id_comp) {
    fetch('/verInstructores', {
        method: 'POST',
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({ id_comp })
    })
    .then(res => res.json())
    .then(data => {
        console.log("Respuesta del backend:", data);

        let instructores = data.instructores || [];
        let competencia = data.competencia || "Sin competencia";

        let competenciaHtml = `<h4 style="margin-top:10px; font-weight:normal">${competencia}</h4>`;

        let buscador = `
            <input type="text" id="buscadorInstructores" 
                   placeholder="Buscar instructor..." 
                   style="width: 95%; padding:5px; margin-bottom:10px; border:1px solid #ccc; border-radius:5px;">
        `;

        // Parámetros de paginación
        let currentPage = 1;
        let rowsPerPage = 10; // 👈 cuántos registros por página
        let totalPages = Math.ceil(instructores.length / rowsPerPage);

        function renderTable(page = 1, filter = "") {
            let start = (page - 1) * rowsPerPage;
            let end = start + rowsPerPage;

            let filtered = instructores.filter(ins => 
                ins.nombre.toLowerCase().includes(filter) ||
                (ins.perfil || "").toLowerCase().includes(filter) ||
                (ins.rap || "").toLowerCase().includes(filter)
            );

            totalPages = Math.ceil(filtered.length / rowsPerPage);

            let tabla = `
            <style>
                #tablaInstructores { border-collapse: collapse; width: 100%; text-align: left; }
                #tablaInstructores th, #tablaInstructores td { padding: 8px 12px; border: 1px solid #ddd; }
                #tablaInstructores th:nth-child(1), #tablaInstructores td:nth-child(1) { width: 170px; }
                #tablaInstructores th:nth-child(2), #tablaInstructores td:nth-child(2) { width: 170px; }
                #tablaInstructores th:nth-child(3), #tablaInstructores td:nth-child(3) { width: 290px; }
                #tablaInstructores th:nth-child(4), #tablaInstructores td:nth-child(4) { width: 60px; text-align: center; }
                .pagination { margin-top: 10px; text-align: center; }
                .pagination button { margin: 0 5px; padding: 5px 10px; }
            </style>
            <table id="tablaInstructores" class="table table-bordered">
                <thead>
                <tr>
                    <th>Instructor</th>
                    <th>Perfil</th>
                    <th>RAP</th>
                    <th>Opciones</th>
                </tr>
                </thead>
                <tbody>
            `;

            filtered.slice(start, end).forEach(ins => {
                tabla += `
                    <tr>
                        <td>${ins.nombre}</td>
                        <td>${ins.perfil || 'Sin perfil'}</td>
                        <td>${ins.rap || 'Sin RAP'}</td>
                        <td>
                            <button onclick="eliminarAsociaciones(${id_comp}, ${ins.id_instructor})" 
                                    class="btn btn-sm btn-danger">
                                Eliminar
                            </button>
                        </td>
                    </tr>
                `;
            });

            tabla += `</tbody></table>`;

            // Paginación
            tabla += `
                <div class="pagination">
                    <button id="prevPage" ${page === 1 ? "disabled" : ""}>Anterior</button>
                    <span>Página ${page} de ${totalPages}</span>
                    <button id="nextPage" ${page === totalPages ? "disabled" : ""}>Siguiente</button>
                </div>
            `;

            return competenciaHtml + buscador + tabla;
        }

        // Render inicial
        Swal.fire({
            title: 'Instructores Asociados',
            html: renderTable(currentPage),
            width: 1500,
            showConfirmButton: false,
            showCloseButton: true,
            background: "#fff",
            backdrop: "rgba(0,0,0,0.4)",
            didOpen: () => {
                const input = document.getElementById("buscadorInstructores");
                input.addEventListener("keyup", function() {
                    let filter = input.value.toLowerCase();
                    Swal.update({ html: renderTable(1, filter) }); // Reinicia búsqueda en página 1
                });

                document.addEventListener("click", function(e) {
                    if (e.target && e.target.id === "prevPage" && currentPage > 1) {
                        currentPage--;
                        Swal.update({ html: renderTable(currentPage, input.value.toLowerCase()) });
                    }
                    if (e.target && e.target.id === "nextPage" && currentPage < totalPages) {
                        currentPage++;
                        Swal.update({ html: renderTable(currentPage, input.value.toLowerCase()) });
                    }
                });
            }
        });
    })
    .catch(err => {
        console.error("Error en verInstructores:", err);
        Swal.fire("Error", "No se pudieron cargar los instructores.", "error");
    });
}




    function eliminarAsociaciones(id_comp, id_instructor) {
    fetch('/eliminarAsociaciones', {
        method: 'POST',
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({ id_comp, id_instructor })
    })
    .then(res => res.json())
    .then(data => {
        if (data.success) {
        verInstructores(id_comp);
        } else {
        Swal.fire({ icon: 'error', text: 'No se pudo eliminar.', didOpen: customSwalStyling });
        }
    });
    }



    function editarCompetencia(id, btn) {
      const row = btn.closest('tr');
      const tipoActual = row.querySelector('.tipo').innerText.trim() === 'Transversal' ? '0' : '1';
      const sigla = row.querySelector('.sigla').innerText;
    
      Swal.fire({
        title: 'Editar Competencia',
        html: `
          <select id="tipo" class="swal2-select">
            <option value="0" ${tipoActual === '0' ? 'selected' : ''}>Transversal</option>
            <option value="1" ${tipoActual === '1' ? 'selected' : ''}>Técnica</option>
          </select>
          <input id="sigla" class="swal2-input" placeholder="Sigla" value="${sigla}">
        `,
        confirmButtonText: 'Guardar',
        showCancelButton: true,
        showCloseButton: true,
        confirmButtonColor: "#1e7e34",
        background: "#fff",
        backdrop: "rgba(0,0,0,0.4)",
        didOpen: customSwalStyling,
        preConfirm: () => {
          const valorTipo = document.getElementById('tipo').value;  // ya es "0" o "1"
          const nuevaSigla = document.getElementById('sigla').value;
          return fetch('/editarComp', {
            method: 'POST',
            headers: {'Content-Type': 'application/json'},
            body: JSON.stringify({ id, tipo: valorTipo, sigla: nuevaSigla })
          })
          .then(res => res.json())
          .then(data => {
            if (data.success) {
              row.querySelector('.tipo').innerText = data.tipo == 0 ? 'Transversal' : 'Técnica';
              row.querySelector('.sigla').innerText = data.sigla;
              Swal.fire({ icon:'success', text:'Actualizado correctamente!', didOpen: customSwalStyling,  confirmButtonColor: '#1e7e34' });
            }
          });
        }
      });
    }


    function verCompetencia(id) {
      fetch('/verComp', {
        method: 'POST',
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({id: id})
      })
      .then(res => res.json())
      .then(data => {
        let html = `<strong>Sigla:</strong> ${data.sigla}<br><br>`;

        // Agregar botones para gestión masiva
        html += `
          <div class="mb-3">
            <button onclick="agregarPerfilMasivo(${id})" class="btn btn-sm btn-success me-2">
              <i class="fa fa-plus"></i> Agregar Perfil a Múltiples RAPs
            </button>
            <button onclick="eliminarPerfilMasivo(${id})" class="btn btn-sm btn-warning">
              <i class="fa fa-minus"></i> Eliminar Perfil de Múltiples RAPs
            </button>
          </div>
          <hr>
        `;

        data.raps.forEach(rap => {
          html += `
            <div style="border:1px solid #ccc; padding:10px; margin-bottom:10px">
              <b>${rap.descripcion}</b><br>
              <button onclick="togglePerfiles(${rap.id_raps})" class="btn btn-sm btn-outline-info mt-2">Mostrar Perfiles</button>
              <div id="perfiles-${rap.id_raps}" class="mt-2" style="display: none"></div>
            </div>
          `;
        });

        Swal.fire({
          title: data.nombre,
          showCloseButton: true,
          html: html,
          width: 800,
          confirmButtonColor: "#1e7e34",
          background: "#fff",
          backdrop: "rgba(0,0,0,0.4)",
          didOpen: customSwalStyling
        });
      });
    }


    function togglePerfiles(id_raps) {
      const cont = document.getElementById(`perfiles-${id_raps}`);
      if (cont.style.display === "none") {
        verPerfiles(id_raps);
        cont.style.display = "block";
      } else {
        cont.style.display = "none";
      }
    }

    function verPerfiles(id_raps) {
      fetch('/perfilesRap', {
        method: 'POST',
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({id_raps: id_raps})
      })
      .then(res => res.json())
      .then(data => {
        const cont = document.getElementById(`perfiles-${id_raps}`);
        cont.innerHTML = '';
        data.perfiles.forEach(p => {
          cont.innerHTML += `<div class="d-flex justify-content-between">
            <span>${p.nombre}</span>
            <button onclick="eliminarPerfilRap(${id_raps}, ${p.id})" class="btn btn-sm btn-danger">Eliminar</button>
          </div>`;
        });
        cont.innerHTML += `
          <select id="nuevoPerfil-${id_raps}" class="form-select mt-2">
            ${data.todos.map(p => `<option value="${p.id}">${p.nombre}</option>`).join('')}
          </select>
          <button onclick="agregarPerfilRap(${id_raps})" class="btn btn-sm btn-success mt-1">Agregar</button>
        `;
      });
    }

    function agregarPerfilRap(id_raps) {
      const id_perfil = document.getElementById(`nuevoPerfil-${id_raps}`).value;
      fetch('/guardarPerfilRap', {
        method: 'POST',
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({id_raps: id_raps, id_perfil: id_perfil})
      }).then(res => res.json()).then(data => {
        if (data.success) {
          verPerfiles(id_raps);
        } else {
          Swal.fire({icon: 'error', text: 'Ya está asociado.', didOpen: customSwalStyling,  confirmButtonColor: '#1e7e34'});
        }
      });
    }

    function eliminarPerfilRap(id_raps, id_perfil) {
      fetch('/eliminarPerfilRap', {
        method: 'POST',
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({id_raps: id_raps, id_perfil: id_perfil})
      }).then(() => verPerfiles(id_raps));
    }

    function customSwalStyling() {
      const popup = Swal.getPopup(); 
      const logo = document.createElement("img");
      logo.src = "../../static/img/logo-sena-verde-png-sin-fondo.webp";
      Object.assign(logo.style, {
        position: "absolute",
        top: "50%",
        left: "50%",
        transform: "translate(-50%, -50%)",
        opacity: "0.2",
        width: "220px",
        pointerEvents: "none",
        zIndex: "0"
      });
      popup.insertBefore(logo, popup.firstChild);
      popup.querySelectorAll(".swal2-title, .swal2-html-container, .swal2-actions")
        .forEach(el => el.style.zIndex = "1");
      const successIcon = popup.querySelector(".swal2-icon.swal2-success");
      if (successIcon) {
        popup.querySelectorAll(".swal2-success-circular-line-left, .swal2-success-circular-line-right")
          .forEach(el => el.style.backgroundColor = "transparent");
        const fix = popup.querySelector(".swal2-success-fix");
        if (fix) fix.style.backgroundColor = "transparent";
        const ring = popup.querySelector(".swal2-success-ring");
        if (ring) ring.style.zIndex = "1";
        popup.querySelectorAll(".swal2-success-line-tip, .swal2-success-line-long")
          .forEach(el => el.style.zIndex = "2");
      }
    }

    document.getElementById('buscador').addEventListener('keyup', function() {
      const valor = this.value.toLowerCase();
      document.querySelectorAll('#tablaCompetencias tbody tr').forEach(row => {
        const texto = row.innerText.toLowerCase();
        row.style.display = texto.includes(valor) ? '' : 'none';
      });
    });




// --- AGREGAR MASIVO ---
function agregarPerfilMasivo(id_comp) {
  Promise.all([
    fetch('/verComp', {
      method: 'POST',
      headers: {'Content-Type': 'application/json'},
      body: JSON.stringify({id: id_comp})
    }).then(r => r.json()),
    fetch('/perfilesRap', {
      method: 'POST',
      headers: {'Content-Type': 'application/json'},
      body: JSON.stringify({id_comp: id_comp})
    }).then(r => r.json())
  ]).then(([compData, perfilData]) => {
    const raps = compData.raps || [];
    const total_raps = perfilData.total_raps || 0;
    const perfiles_status = perfilData.perfiles_status || [];

    // Mostrar solo perfiles que NO están en todos los RAPs
    const perfilesDisponibles = perfiles_status.filter(p => p.associated_count < total_raps);

    if (perfilesDisponibles.length === 0) {
      Swal.fire({ icon: 'info', text: 'No hay perfiles disponibles para agregar. Todos los perfiles ya están asociados a todos los RAPs.', didOpen: customSwalStyling });
      return;
    }

    const html = `
      <div class="mb-3">
        <label class="form-label"><strong>Seleccionar Perfil:</strong></label>
        <select id="perfilMasivo" class="form-select">
          ${perfilesDisponibles.map(p => `<option value="${p.id}">${p.nombre}</option>`).join('')}
        </select>
      </div>
      <div class="mb-3">
        <label class="form-label"><strong>Seleccionar RAPs:</strong></label>
        <div id="rapsContainer" style="max-height:200px; overflow-y:auto; border:1px solid #ddd; padding:10px;">
          <!-- Se cargarán los RAPs faltantes aquí -->
        </div>
      </div>
      <div class="mb-2">
        <button id="selTodosAgregar" class="btn btn-sm btn-outline-primary me-2">Seleccionar Todos</button>
        <button id="deselTodosAgregar" class="btn btn-sm btn-outline-secondary">Deseleccionar Todos</button>
      </div>
    `;

    Swal.fire({
      title: 'Agregar Perfil a Múltiples RAPs',
      html,
      width: 650,
      showCancelButton: true,
      confirmButtonText: 'Agregar',
      cancelButtonText: 'Cancelar',
      confirmButtonColor: "#1e7e34",
      showCloseButton: true,
      didOpen: () => {
        customSwalStyling();

        const perfilSelect = document.getElementById('perfilMasivo');
        const rapsContainer = document.getElementById('rapsContainer');
        const selTodosBtn = document.getElementById('selTodosAgregar');
        const deselTodosBtn = document.getElementById('deselTodosAgregar');

        // Carga los RAPs faltantes para un perfil usando /rapsPorPerfil
        function cargarRapsFaltantes(id_perfil) {
          rapsContainer.innerHTML = `<div class="text-muted">Cargando...</div>`;
          fetch('/rapsPorPerfil', {
            method: 'POST',
            headers: {'Content-Type': 'application/json'},
            body: JSON.stringify({ id_comp: id_comp, id_perfil: id_perfil })
          })
          .then(r => r.json())
          .then(data => {
            const faltantes = data.raps_faltantes || [];
            if (!faltantes.length) {
              rapsContainer.innerHTML = `<div class="text-muted">Este perfil ya está asociado a todos los RAPs de la competencia.</div>`;
              return;
            }
            rapsContainer.innerHTML = faltantes.map(rap => `
              <div class="form-check">
                <input class="form-check-input" type="checkbox" value="${rap.id_raps}" id="rap-${rap.id_raps}">
                <label class="form-check-label" for="rap-${rap.id_raps}">${rap.descripcion}</label>
              </div>
            `).join('');
          })
          .catch(err => {
            console.error(err);
            rapsContainer.innerHTML = `<div class="text-danger">Error al cargar RAPs.</div>`;
          });
        }

        // Inicial
        if (perfilSelect.value) cargarRapsFaltantes(perfilSelect.value);

        // Al cambiar perfil, recargar faltantes
        perfilSelect.addEventListener('change', function(){ cargarRapsFaltantes(this.value); });

        selTodosBtn.addEventListener('click', () => document.querySelectorAll('#rapsContainer input[type="checkbox"]').forEach(cb => cb.checked = true));
        deselTodosBtn.addEventListener('click', () => document.querySelectorAll('#rapsContainer input[type="checkbox"]').forEach(cb => cb.checked = false));
      },
      preConfirm: () => {
        const id_perfil = document.getElementById('perfilMasivo').value;
        const rapsSeleccionados = Array.from(document.querySelectorAll('#rapsContainer input[type="checkbox"]:checked')).map(cb => cb.value);
        if (rapsSeleccionados.length === 0) {
          Swal.showValidationMessage('Debes seleccionar al menos un RAP faltante para agregar.');
          return false;
        }
        return procesarPerfilMasivo(id_perfil, rapsSeleccionados, 'agregar');
      }
    });

  }).catch(err => {
    console.error(err);
    Swal.fire('Error', 'No se pudo cargar la información.', 'error');
  });
}


// --- ELIMINAR MASIVO ---
function eliminarPerfilMasivo(id_comp) {
  Promise.all([
    fetch('/verComp', {
      method: 'POST',
      headers: {'Content-Type': 'application/json'},
      body: JSON.stringify({id: id_comp})
    }).then(r => r.json()),
    fetch('/perfilesRap', {
      method: 'POST',
      headers: {'Content-Type': 'application/json'},
      body: JSON.stringify({id_comp: id_comp})
    }).then(r => r.json())
  ]).then(([compData, perfilData]) => {
    const raps = compData.raps || [];
    const perfiles_status = perfilData.perfiles_status || [];

    // Mostrar solo perfiles que tengan al menos 1 asociación dentro de la competencia
    const perfilesDisponibles = perfiles_status.filter(p => (p.associated_count || 0) > 0);

    if (perfilesDisponibles.length === 0) {
      Swal.fire({ icon: 'info', text: 'No hay perfiles asociados a esta competencia.', didOpen: customSwalStyling,confirmButtonColor: '#1e7e34' });
      return;
    }

    const html = `
      <div class="mb-3">
        <label class="form-label"><strong>Seleccionar Perfil:</strong></label>
        <select id="perfilMasivoEliminar" class="form-select">
          ${perfilesDisponibles.map(p => `<option value="${p.id}">${p.nombre}</option>`).join('')}
        </select>
      </div>
      <div class="mb-3">
        <label class="form-label"><strong>Seleccionar RAPs:</strong></label>
        <div id="rapsEliminarContainer" style="max-height:200px; overflow-y:auto; border:1px solid #ddd; padding:10px;">
          <!-- Se cargarán los RAPs asociados aquí -->
        </div>
      </div>
      <div class="mb-2">
        <button id="selTodosEliminar" class="btn btn-sm btn-outline-primary me-2">Seleccionar Todos</button>
        <button id="deselTodosEliminar" class="btn btn-sm btn-outline-secondary">Deseleccionar Todos</button>
      </div>
    `;

    Swal.fire({
      title: 'Eliminar Perfil de Múltiples RAPs',
      html,
      width: 650,
      showCancelButton: true,
      confirmButtonText: 'Eliminar',
      cancelButtonText: 'Cancelar',
      confirmButtonColor: "#dc3545",
      showCloseButton: true,
      didOpen: () => {
        customSwalStyling();

        const perfilSelect = document.getElementById('perfilMasivoEliminar');
        const rapsEliminarContainer = document.getElementById('rapsEliminarContainer');
        const selTodosBtn = document.getElementById('selTodosEliminar');
        const deselTodosBtn = document.getElementById('deselTodosEliminar');

        function cargarRapsAsociados(id_perfil) {
          rapsEliminarContainer.innerHTML = `<div class="text-muted">Cargando...</div>`;
          fetch('/rapsPorPerfil', {
            method: 'POST',
            headers: {'Content-Type': 'application/json'},
            body: JSON.stringify({ id_comp: id_comp, id_perfil: id_perfil })
          })
          .then(r => r.json())
          .then(data => {
            const asociados = data.raps_asociados || [];
            if (!asociados.length) {
              rapsEliminarContainer.innerHTML = `<div class="text-muted">Este perfil no está asociado a ningún RAP de la competencia.</div>`;
              return;
            }
            rapsEliminarContainer.innerHTML = asociados.map(rap => `
              <div class="form-check">
                <input class="form-check-input" type="checkbox" value="${rap.id_raps}" id="rap-eliminar-${rap.id_raps}">
                <label class="form-check-label" for="rap-eliminar-${rap.id_raps}">${rap.descripcion}</label>
              </div>
            `).join('');
          })
          .catch(err => {
            console.error(err);
            rapsEliminarContainer.innerHTML = `<div class="text-danger">Error al cargar RAPs.</div>`;
          });
        }

        // Inicial
        if (perfilSelect.value) cargarRapsAsociados(perfilSelect.value);
        perfilSelect.addEventListener('change', function(){ cargarRapsAsociados(this.value); });

        selTodosBtn.addEventListener('click', () => document.querySelectorAll('#rapsEliminarContainer input[type="checkbox"]').forEach(cb => cb.checked = true));
        deselTodosBtn.addEventListener('click', () => document.querySelectorAll('#rapsEliminarContainer input[type="checkbox"]').forEach(cb => cb.checked = false));
      },
      preConfirm: () => {
        const id_perfil = document.getElementById('perfilMasivoEliminar').value;
        const rapsSeleccionados = Array.from(document.querySelectorAll('#rapsEliminarContainer input[type="checkbox"]:checked')).map(cb => cb.value);
        if (rapsSeleccionados.length === 0) {
          Swal.showValidationMessage('Debes seleccionar al menos un RAP asociado para eliminar.');
          return false;
        }
        return procesarPerfilMasivo(id_perfil, rapsSeleccionados, 'eliminar');
      }
    });

  }).catch(err => {
    console.error(err);
    Swal.fire('Error', 'No se pudo cargar la información.', 'error');
  });
}


function seleccionarTodosRaps(seleccionar) {
  const checkboxes = document.querySelectorAll('input[id^="rap-"]:not([id*="eliminar"])');
  checkboxes.forEach(cb => cb.checked = seleccionar);
}

function seleccionarTodosRapsEliminar(seleccionar) {
  const checkboxes = document.querySelectorAll('input[id^="rap-eliminar-"]');
  checkboxes.forEach(cb => cb.checked = seleccionar);
}

function procesarPerfilMasivo(id_perfil, rapsSeleccionados, accion) {
  const promesas = rapsSeleccionados.map(id_raps => {
    if (accion === 'agregar') {
      return fetch('/guardarPerfilRap', {
        method: 'POST',
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({id_raps: id_raps, id_perfil: id_perfil})
      }).then(res => res.json());
    } else {
      return fetch('/eliminarPerfilRap', {
        method: 'POST',
        headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({id_raps: id_raps, id_perfil: id_perfil})
      }).then(res => res.json());
    }
  });
  
  return Promise.all(promesas).then(resultados => {
    const exitosos = resultados.filter(r => r.success !== false).length;
    const fallidos = resultados.length - exitosos;
    
    let mensaje = '';
    if (accion === 'agregar') {
      mensaje = `Perfil agregado a ${exitosos} RAPs`;
      if (fallidos > 0) {
        mensaje += ` (${fallidos} ya estaban asociados)`;
      }
    } else {
      mensaje = `Perfil eliminado de ${exitosos} RAPs`;
    }
    
    Swal.fire({
      icon: 'success',
      title: 'Proceso completado',
      text: mensaje,
      confirmButtonColor: '#1e7e34',
      didOpen: customSwalStyling
    });
  }).catch(error => {
    Swal.fire({
      icon: 'error',
      title: 'Error',
      text: 'Hubo un problema al procesar la operación',
      confirmButtonColor: '#1e7e34',
      didOpen: customSwalStyling
    });
  });
}

File: Src/routes/perfilesAsociarComp.py
# Flask - dentro del blueprint perfilesAsociarComp_bp
from flask import Blueprint, render_template, request, jsonify
from flask_mysqldb import MySQL

perfilesAsociarComp_bp = Blueprint('perfilesAsociarComp', __name__)
mysql = MySQL()

@perfilesAsociarComp_bp.route("/perfilesAsociarComp")
def perfilesAsociarComp():
    cur = mysql.connection.cursor()
    query = """
        SELECT c.id_comp, c.nombre, c.tipo, c.sigla, p.nombre AS programa
        FROM competencias c
        LEFT JOIN programas p ON c.id_prog = p.id_prog
    """
    cur.execute(query)
    rows = cur.fetchall()
    competencias = [
        {
            'id': row[0],
            'nombre': row[1],
            'tipo': row[2] if row[2] is not None else '',
            'sigla': row[3],
            'programa': row[4]
        }
        for row in rows
    ]
    return render_template('perfilesAsociarComp.html', competencias=competencias)

@perfilesAsociarComp_bp.route("/editarComp", methods=['POST'])
def editarComp():
    data = request.get_json()
    id_comp = data['id']
    tipo = data['tipo']
    sigla = data['sigla']
    cur = mysql.connection.cursor()
    cur.execute("UPDATE competencias SET tipo=%s, sigla=%s WHERE id_comp=%s", (tipo, sigla, id_comp))
    mysql.connection.commit()
    return jsonify(success=True, tipo=tipo, sigla=sigla)

@perfilesAsociarComp_bp.route("/verComp", methods=['POST'])
def verComp():
    data = request.get_json()
    id_comp = data['id']
    cur = mysql.connection.cursor()
    cur.execute("SELECT nombre, sigla FROM competencias WHERE id_comp = %s", (id_comp,))
    nombre, sigla = cur.fetchone()
    cur.execute("SELECT id_raps, descripcion FROM raps WHERE id_comp = %s", (id_comp,))
    raps = [{'id_raps': r[0], 'descripcion': r[1]} for r in cur.fetchall()]
    return jsonify(nombre=nombre, sigla=sigla, raps=raps)

@perfilesAsociarComp_bp.route("/perfilesRap", methods=['POST'])
def perfilesRap():
    data = request.get_json()
    cur = mysql.connection.cursor()

    # Si llega id_raps: comportamiento antiguo (por compatibilidad)
    if data.get('id_raps') is not None:
        id_raps = data['id_raps']
        cur.execute("""
            SELECT p.id_perfil, p.nombre FROM perfiles p
            JOIN perfiles_raps pr ON p.id_perfil = pr.id_perfil
            WHERE pr.id_raps = %s
        """, (id_raps,))
        perfiles_asociados = [{'id': row[0], 'nombre': row[1]} for row in cur.fetchall()]

        cur.execute("SELECT id_perfil, nombre FROM perfiles")
        todos = [{'id': row[0], 'nombre': row[1]} for row in cur.fetchall()]

        return jsonify(perfiles=perfiles_asociados, todos=todos)

    # Si llega id_comp: devolver estado por competencia (para el modal masivo)
    if data.get('id_comp') is not None:
        id_comp = data['id_comp']

        # Obtener todos los RAPs de la competencia
        cur.execute("SELECT id_raps FROM raps WHERE id_comp = %s", (id_comp,))
        rap_rows = [r[0] for r in cur.fetchall()]

        total_raps = len(rap_rows)
        if total_raps == 0:
            return jsonify(perfiles_status=[], total_raps=0)

        # Preparar placeholders para IN (...)
        placeholders = ','.join(['%s'] * total_raps)

        # Para cada perfil, traer los id_raps asociados (solo dentro de los RAPs de esa competencia)
        # y contar cuántos RAPs (dentro de esa competencia) tiene asociado.
        cur.execute(f"""
            SELECT p.id_perfil, p.nombre,
                   GROUP_CONCAT(pr.id_raps) AS rap_ids,
                   COUNT(pr.id_raps) AS associated_count
            FROM perfiles p
            LEFT JOIN perfiles_raps pr
              ON p.id_perfil = pr.id_perfil AND pr.id_raps IN ({placeholders})
            GROUP BY p.id_perfil, p.nombre
        """, tuple(rap_rows))

        perfiles_status = []
        for row in cur.fetchall():
            rap_ids_raw = row[2]  # puede ser None
            rap_ids = rap_ids_raw.split(',') if rap_ids_raw else []
            perfiles_status.append({
                'id': row[0],
                'nombre': row[1],
                'associated_count': int(row[3]),
                'rap_ids': rap_ids   # lista de id_raps en string
            })

        return jsonify(perfiles_status=perfiles_status, total_raps=total_raps)

    # Si no se envía nada válido:
    return jsonify(error="Faltan parámetros (id_raps o id_comp)"), 400


@perfilesAsociarComp_bp.route("/guardarPerfilRap", methods=['POST'])
def guardarPerfilRap():
    data = request.get_json()
    id_raps = data['id_raps']
    id_perfil = data['id_perfil']
    cur = mysql.connection.cursor()
    cur.execute("SELECT 1 FROM perfiles_raps WHERE id_raps = %s AND id_perfil = %s", (id_raps, id_perfil))
    if cur.fetchone():
        return jsonify(success=False)
    cur.execute("INSERT INTO perfiles_raps (id_raps, id_perfil) VALUES (%s, %s)", (id_raps, id_perfil))
    mysql.connection.commit()
    return jsonify(success=True)

@perfilesAsociarComp_bp.route("/eliminarPerfilRap", methods=['POST'])
def eliminarPerfilRap():
    data = request.get_json()
    id_raps = data['id_raps']
    id_perfil = data['id_perfil']
    cur = mysql.connection.cursor()
    cur.execute("DELETE FROM perfiles_raps WHERE id_raps = %s AND id_perfil = %s", (id_raps, id_perfil))
    mysql.connection.commit()
    return jsonify(success=True)


@perfilesAsociarComp_bp.route("/verInstructores", methods=['POST']) 
def verInstructores(): 
    data = request.get_json() 
    id_comp = data.get('id_comp') 
    try: 
        cur = mysql.connection.cursor() 
 
        cur.execute(""" 
            SELECT  
                i.id_instructor, 
                db.nombre AS instructor, 
                GROUP_CONCAT(DISTINCT p.nombre SEPARATOR ', ') AS perfiles, 
                r.descripcion AS rap, 
                c.nombre AS competencia 
            FROM asociaciones a 
            JOIN instructores i  
                ON a.id_instructor = i.id_instructor 
            JOIN datos_basicos db  
                ON i.cedula = db.cedula 
            LEFT JOIN perfil_instructor pi  
                ON pi.id_instructor = i.id_instructor 
            LEFT JOIN perfiles p  
                ON p.id_perfil = pi.id_perfil 
            LEFT JOIN raps r  
                ON r.id_comp = a.id_comp 
            JOIN competencias c  
                ON c.id_comp = a.id_comp 
            WHERE a.id_comp = %s 
            GROUP BY i.id_instructor, db.nombre, r.descripcion, c.nombre 
        """, (id_comp,)) 
 
        rows = cur.fetchall() 
 
        instructores = [ 
            { 
                'id_instructor': row[0], 
                'nombre': row[1], 
                'perfil': row[2] if row[2] else 'Sin perfil', 
                'rap': row[3] if row[3] else 'Sin RAP' 
            } 
            for row in rows 
        ] 
 
        competencia = rows[0][4] if rows else "Sin competencia" 
 
        return jsonify(instructores=instructores, competencia=competencia) 
 
    except Exception as e: 
        print("Error en verInstructores:", e) 
        return jsonify(error=str(e)), 500




@perfilesAsociarComp_bp.route("/eliminarAsociaciones", methods=['POST'])
def eliminarAsociaciones():
    data = request.get_json()
    id_comp = data['id_comp']
    id_instructor = data['id_instructor']
    cur = mysql.connection.cursor()
    cur.execute("DELETE FROM asociaciones WHERE id_comp = %s AND id_instructor = %s", (id_comp, id_instructor))
    mysql.connection.commit()
    return jsonify(success=True)



@perfilesAsociarComp_bp.route("/rapsPorPerfil", methods=['POST'])
def rapsPorPerfil():
    data = request.get_json()
    id_comp = data.get('id_comp')
    id_perfil = data.get('id_perfil')

    if id_comp is None or id_perfil is None:
        return jsonify(error="Faltan parámetros (id_comp o id_perfil)"), 400

    cur = mysql.connection.cursor()

    # RAPs asociados al perfil (dentro de la competencia)
    cur.execute("""
        SELECT r.id_raps, r.descripcion
        FROM raps r
        JOIN perfiles_raps pr ON r.id_raps = pr.id_raps
        WHERE r.id_comp = %s AND pr.id_perfil = %s
    """, (id_comp, id_perfil))
    raps_asociados = [{'id_raps': row[0], 'descripcion': row[1]} for row in cur.fetchall()]

    # RAPs de la competencia que NO están asociados al perfil (faltantes)
    cur.execute("""
        SELECT r.id_raps, r.descripcion
        FROM raps r
        WHERE r.id_comp = %s
        AND r.id_raps NOT IN (
            SELECT pr.id_raps FROM perfiles_raps pr WHERE pr.id_perfil = %s
        )
    """, (id_comp, id_perfil))
    raps_faltantes = [{'id_raps': row[0], 'descripcion': row[1]} for row in cur.fetchall()]

    return jsonify(raps_asociados=raps_asociados, raps_faltantes=raps_faltantes)
